File: web/network.py
from .nanowebapi import HttpError
from .webserver import read_json, CREDENTIALS, authenticate
from lib.kernel import Service
import uasyncio as asyncio
import network
import ubinascii
import json
import time


class NetworkApi(Service):
    def __init__(self, name, web):
        super().__init__(name)
        self.web = web
        web.web_services.append(self.__class__.__name__)

        self.web.app.route('/api/network/status')(self.api_status)
        self.web.app.route('/api/network/scan')(self.api_scan)
        self.web.app.route('/api/network/save')(self.api_save)

    @authenticate(CREDENTIALS)
    async def api_status(self, request):
        if request.method == "OPTIONS": return await self.web.api_send_response(request)

        sta = network.WLAN(network.STA_IF)
        ap = network.WLAN(network.AP_IF)

        curr_time = time.localtime()
        time_synced = curr_time[0] > 2022

        sta_ip, sta_mask, sta_gw, sta_dns = sta.ifconfig() if sta.isconnected() else ("0.0.0.0", "0.0.0.0", "0.0.0.0",
                                                                                      "0.0.0.0")

        try:
            ap_ssid_current = ap.config('essid') if ap.active() else ""
        except:
            ap_ssid_current = ""

        status = {
            "sta_connected": sta.isconnected(),
            "sta_ip": sta_ip,
            "sta_mask": sta_mask,
            "sta_gw": sta_gw,
            "sta_dns": sta_dns,
            "sta_rssi": 0,
            "ap_active": ap.active(),
            "ap_ip": ap.ifconfig()[0] if ap.active() else "0.0.0.0",
            "ap_ssid_current": ap_ssid_current,
            "time_synced": time_synced,
            "current_time": "{:02}:{:02}:{:02}".format(curr_time[3], curr_time[4], curr_time[5])
        }

        if sta.isconnected():
            try:
                status["sta_rssi"] = sta.status('rssi')
            except:
                pass

        try:
            with open('wifi.json', 'r') as f:
                config = json.load(f)
            if config.get('sta_pass'): config['sta_pass'] = '********'
            if config.get('ap_pass'): config['ap_pass'] = '********'
            status['config'] = config
        except:
            status['config'] = {}

        await self.web.api_send_response(request, data=status)

    @authenticate(CREDENTIALS)
    async def api_scan(self, request):
        if request.method == "OPTIONS": return await self.web.api_send_response(request)
        sta = network.WLAN(network.STA_IF)
        if not sta.active(): sta.active(True)
        try:
            nets = sta.scan()
            nets.sort(key=lambda x: x[3], reverse=True)
            results = []
            for n in nets:
                try:
                    ssid = n[0].decode('utf-8')
                    if ssid: results.append({"ssid": ssid, "rssi": n[3]})
                except:
                    pass
            await self.web.api_send_response(request, data=results)
        except Exception:
            raise HttpError(request, 500, "Scan failed")

    @authenticate(CREDENTIALS)
    async def api_save(self, request):
        if request.method == "OPTIONS": return await self.web.api_send_response(request)

        data = await read_json(request)
        save_type = data.get('save_type', 'all')  # 'sta' или 'ap'

        try:
            with open('wifi.json', 'r') as f:
                config = json.load(f)
        except:
            config = {}

        # 1. Если сохраняем РОУТЕР
        if save_type == 'sta':
            new_pass = data.get('sta_pass', '')
            if new_pass == "" or new_pass == "********":
                new_pass = config.get('sta_pass', '')

            config['sta_ssid'] = data.get('sta_ssid', '')
            config['sta_pass'] = new_pass
            config['sta_static'] = data.get('sta_static', False)
            config['sta_ip'] = data.get('sta_ip', '')
            config['sta_mask'] = data.get('sta_mask', '')
            config['sta_gw'] = data.get('sta_gw', '')
            config['sta_dns'] = data.get('sta_dns', '')

        # 2. Если сохраняем ТОЧКУ ДОСТУПА
        elif save_type == 'ap':
            ap = network.WLAN(network.AP_IF)
            mac_suffix = ubinascii.hexlify(ap.config('mac'))[-4:].decode('utf-8').upper()
            default_ap_name = f"HLOS_{mac_suffix}"

            new_ap_ssid = data.get('ap_ssid', '').strip()
            if not new_ap_ssid:
                new_ap_ssid = config.get('ap_ssid', default_ap_name)

            new_ap_pass = data.get('ap_pass', '')
            if new_ap_pass == "" or new_ap_pass == "********":
                new_ap_pass = config.get('ap_pass', '')
            elif 0 < len(new_ap_pass) < 8:
                new_ap_pass = config.get('ap_pass', '')

            config['ap_ssid'] = new_ap_ssid
            config['ap_pass'] = new_ap_pass
            config['ap_disable'] = data.get('ap_disable', False)

        with open('wifi.json', 'w') as f:
            json.dump(config, f)

        await self.web.api_send_response(request, data={"status": "trying_to_connect"})

        async def apply_settings():
            await asyncio.sleep(1)

            if save_type == 'sta':
                sta = network.WLAN(network.STA_IF)
                sta.active(True)
                sta.disconnect()
                await asyncio.sleep(0.5)
                if config.get('sta_static') and config.get('sta_ip'):
                    ip = config.get('sta_ip')
                    mask = config.get('sta_mask') if config.get('sta_mask') else '255.255.255.0'
                    gw = config.get('sta_gw') if config.get('sta_gw') else ip
                    dns = config.get('sta_dns') if config.get('sta_dns') else '8.8.8.8'
                    try:
                        sta.ifconfig((ip, mask, gw, dns))
                    except Exception as e:
                        sta.ifconfig('dhcp')
                else:
                    sta.ifconfig('dhcp')
                sta.connect(config['sta_ssid'], config.get('sta_pass', ''))

            elif save_type == 'ap':
                ap = network.WLAN(network.AP_IF)
                # ЖЕСТКИЙ ПЕРЕЗАПУСК AP, чтобы пароль точно применился
                ap.active(False)
                await asyncio.sleep(0.5)
                ap.active(True)
                try:
                    if config.get('ap_pass') and len(config['ap_pass']) >= 8:
                        ap.config(essid=config['ap_ssid'], password=config['ap_pass'], authmode=3)
                    else:
                        ap.config(essid=config['ap_ssid'], authmode=0)
                except Exception:
                    pass

        asyncio.create_task(apply_settings())

File: web/network.js
document.addEventListener('DOMContentLoaded', () => {
    const stConn = document.getElementById('st-conn');
    const stIp = document.getElementById('st-ip');
    const stRssi = document.getElementById('st-rssi');
    const stTime = document.getElementById('st-time');
    const espClock = document.getElementById('esp-clock');
    
    const staSsid = document.getElementById('sta-ssid');
    const staPass = document.getElementById('sta-pass');

    const staStaticCb = document.getElementById('sta-static-cb');
    const staticFields = document.getElementById('static-fields');
    const staIp = document.getElementById('sta-ip');
    const staMask = document.getElementById('sta-mask');
    const staGw = document.getElementById('sta-gw');
    const staDns = document.getElementById('sta-dns');

    const apSsid = document.getElementById('ap-ssid');
    const apPass = document.getElementById('ap-pass');
    const apDisableCb = document.getElementById('ap-disable-cb');

    const scanBtn = document.getElementById('scan-btn');
    const scanRes = document.getElementById('scan-results');

    // ДВЕ КНОПКИ
    const saveStaBtn = document.getElementById('save-sta-btn');
    const saveStaStatus = document.getElementById('save-sta-status');
    const saveApBtn = document.getElementById('save-ap-btn');
    const saveApStatus = document.getElementById('save-ap-status');

    let configLoaded = false;

    staStaticCb.addEventListener('change', () => {
        staticFields.style.display = staStaticCb.checked ? 'block' : 'none';
    });

    async function updateStatus() {
        try {
            const res = await fetch('/api/network/status');
            const data = await res.json();

            if (data.sta_connected) {
                stConn.className = 'status-box ok'; stConn.querySelector('.status-val').textContent = "Подключено";
                stIp.className = 'status-box ok'; stIp.querySelector('.status-val').textContent = data.sta_ip;
                stRssi.querySelector('.status-val').textContent = data.sta_rssi + " dBm";
            } else {
                stConn.className = 'status-box err'; stConn.querySelector('.status-val').textContent = "Отключено";
                stIp.className = 'status-box'; stIp.querySelector('.status-val').textContent = "---";
                stRssi.querySelector('.status-val').textContent = "---";
            }

            stTime.className = data.time_synced ? 'status-box ok' : 'status-box err';
            stTime.querySelector('.status-val').textContent = data.time_synced ? "ОК" : "Нет";
            espClock.textContent = data.current_time;

            if (!configLoaded && data.config) {
                staSsid.value = data.config.sta_ssid || '';

                staPass.value = data.config.sta_pass || '';
                if (data.config.sta_pass) staPass.placeholder = "Пароль сохранен";

                apSsid.value = data.config.ap_ssid || data.ap_ssid_current || '';

                apPass.value = data.config.ap_pass || '';
                if (data.config.ap_pass) apPass.placeholder = "Пароль сохранен";

                staStaticCb.checked = data.config.sta_static === true;
                staStaticCb.dispatchEvent(new Event('change'));

                staIp.value = data.config.sta_ip || (data.sta_ip !== "0.0.0.0" ? data.sta_ip : '');
                staMask.value = data.config.sta_mask || (data.sta_mask !== "0.0.0.0" ? data.sta_mask : '255.255.255.0');
                staGw.value = data.config.sta_gw || (data.sta_gw !== "0.0.0.0" ? data.sta_gw : '');
                staDns.value = data.config.sta_dns || (data.sta_dns !== "0.0.0.0" ? data.sta_dns : '8.8.8.8');

                apDisableCb.checked = data.config.ap_disable === true;
                configLoaded = true;
            }
        } catch (e) { console.error(e); }
    }

    updateStatus();
    setInterval(updateStatus, 3000);

    scanBtn.addEventListener('click', async () => {
        scanBtn.disabled = true; scanBtn.textContent = "Поиск...";
        scanRes.style.display = 'block'; scanRes.innerHTML = '<div style="padding:10px;">Сканирование...</div>';
        try {
            const res = await fetch('/api/network/scan');
            const nets = await res.json();
            scanRes.innerHTML = '';
            if (nets.length === 0) scanRes.innerHTML = '<div style="padding:10px;">Пусто</div>';
            nets.forEach(n => {
                const div = document.createElement('div');
                div.className = 'wifi-item';
                div.innerHTML = `<span>${n.ssid}</span> <span class="wifi-rssi">${n.rssi} dBm</span>`;
                div.onclick = () => { staSsid.value = n.ssid; scanRes.style.display = 'none'; staPass.focus(); };
                scanRes.appendChild(div);
            });
        } catch (e) { scanRes.innerHTML = '<div style="padding:10px; color:red;">Ошибка</div>'; }
        finally { scanBtn.disabled = false; scanBtn.textContent = "🔍 Скан"; }
    });

    // ОБРАБОТЧИК ДЛЯ РОУТЕРА (STA)
    saveStaBtn.addEventListener('click', async () => {
        saveStaBtn.disabled = true;
        saveStaStatus.style.color = "blue";
        saveStaStatus.textContent = "Подключение...";

        const payload = {
            save_type: 'sta', // Флаг для сервера
            sta_ssid: staSsid.value.trim(),
            sta_pass: staPass.value.trim(),
            sta_static: staStaticCb.checked,
            sta_ip: staIp.value.trim(),
            sta_mask: staMask.value.trim(),
            sta_gw: staGw.value.trim(),
            sta_dns: staDns.value.trim()
        };

        try {
            await fetch('/api/network/save', { method: 'POST', headers: {'Content-Type': 'application/json'}, body: JSON.stringify(payload) });
            let attempts = 0;
            const checkConnect = setInterval(() => {
                attempts++;
                if (stConn.querySelector('.status-val').textContent === "Подключено") {
                    clearInterval(checkConnect);
                    saveStaStatus.style.color = "green";
                    saveStaStatus.textContent = "✅ Подключено!";
                    saveStaBtn.disabled = false;
                }
                if (attempts > 15) {
                    clearInterval(checkConnect);
                    saveStaStatus.style.color = "red";
                    saveStaStatus.textContent = "❌ Ошибка.";
                    saveStaBtn.disabled = false;
                }
            }, 3000);
        } catch (e) {
            saveStaStatus.textContent = "Ошибка связи";
            saveStaBtn.disabled = false;
        }
    });

    // ОБРАБОТЧИК ДЛЯ ТОЧКИ ДОСТУПА (AP)
    saveApBtn.addEventListener('click', async () => {
        saveApBtn.disabled = true;
        saveApStatus.style.color = "blue";
        saveApStatus.textContent = "Применяем настройки...";

        const payload = {
            save_type: 'ap', // Флаг для сервера
            ap_ssid: apSsid.value.trim(),
            ap_pass: apPass.value.trim(),
            ap_disable: apDisableCb.checked
        };

        try {
            await fetch('/api/network/save', { method: 'POST', headers: {'Content-Type': 'application/json'}, body: JSON.stringify(payload) });
            setTimeout(() => {
                saveApStatus.style.color = "green";
                saveApStatus.textContent = "✅ Сохранено!";
                saveApBtn.disabled = false;
            }, 2000);
        } catch (e) {
            saveApStatus.textContent = "Ошибка связи";
            saveApBtn.disabled = false;
        }
    });
});